File: kalba-rest-api/src/main/java/kalba/controller/AccountController.java
package kalba.controller;

import kalba.models.account.*;
import kalba.repository.JpaAccountRepository;
import kalba.service.AccountService;
import kalba.service.JwtUserDetailsService;
import kalba.util.JwtTokenUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.DisabledException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.Map;

@RestController
@RequiredArgsConstructor
@RequestMapping("/account")
public class AccountController {
    private static final String NOT_EXIST_USER = "존재하는 아이디입니다.";

    private final AccountService accountService;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtTokenUtil jwtTokenUtil;
    private final JwtUserDetailsService userDetailsService;

    private final JpaAccountRepository jpaAccountRepository;

    @PostMapping("/register")
    public ResponseEntity<Map<Object, Object>> register(@RequestBody RegisterForm registerForm) {
        Account account = Account.builder().nickname(registerForm.getNickname())
                .tag(registerForm.getTag())
                .name(registerForm.getName())
                .password(passwordEncoder.encode(registerForm.getPassword()))
                .role("USER")
                .build();
        int result = accountService.register(account);
        if (result > 0) {
            return ResponseEntity.created(URI.create("/account/" + account.getId())).body(Map.of("message", "success"));
        } else {
            String msg;
            if (result == -1) {
                msg = NOT_EXIST_USER;
            } else if (result == -2) {
                msg = "해당 태그로 가입된 아이디가 존재합니다.";
            } else {
                msg = "예상하지 못한 에러가 발생하였습니다.";
            }
            return new ResponseEntity<>(Map.of("message", msg), HttpStatus.CONFLICT);
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> createAuthenticationToken(@RequestBody JwtRequest request) {
        try {
            authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(request.getUsername(), request.getPassword()));
        } catch (DisabledException e) {
            return new ResponseEntity<>(Map.of("message", "disable user"), HttpStatus.UNAUTHORIZED);
        } catch (BadCredentialsException e) {
            return new ResponseEntity<>(Map.of("message", "invalid credentials"), HttpStatus.UNAUTHORIZED);
        }
        final UserDetails userDetails = userDetailsService.loadUserByUsername(request.getUsername());
        final String token = jwtTokenUtil.generateToken(userDetails);
        return ResponseEntity.ok(new JwtResponse(token));
    }

    @PostMapping("/login/name")
    public ResponseEntity<?> getUserNameFromToken(@RequestBody Token token) {
        return ResponseEntity.ok(Map.of("name", jwtTokenUtil.getUsernameFromToken(token.getToken())));
    }

    @PostMapping("/login/tag")
    public ResponseEntity<?> getTagByName(@RequestBody Name name) {
        return accountService.findAccountByName(name.getName())
                .<ResponseEntity<?>>map(account -> ResponseEntity.ok(Map.of("tag", account.getTag())))
                .orElseGet(() -> ResponseEntity.ok(Map.of("message", "invalid name")));
    }

    @PostMapping("/info")
    public ResponseEntity<?> getAccountInfoByName(@RequestBody Name name) {
        return accountService.findAccountInfoByName(name.getName())
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.ok(Map.of("message", "invalid name")));
    }

    @GetMapping("/info/all")
    public ResponseEntity<?> getAllAccountInfo() {
        return ResponseEntity.ok(accountService.findAllAccountInfo());
    }

    @GetMapping("/admin")
//    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> isAdmin() {
        return ResponseEntity.ok(Map.of("isAdmin", jpaAccountRepository.findById()));
    }
}

File: kalba/src/main/kotlin/kr/kalba/presentation/controller/AccountController.kt
package kr.kalba.presentation.controller

import kr.kalba.application.AccountService
import kr.kalba.application.ClanService
import kr.kalba.presentation.dto.*
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/account")
class AccountController(
    val accountService: AccountService,
    val clanService: ClanService
) {

    @PostMapping("/register")
    fun register(@RequestBody request: RegisterDto.Request): ResponseEntity<RegisterDto.Response> {
        return ResponseEntity.status(HttpStatus.OK).body(
            RegisterDto.Response.of(
                accountService.register(request)
            )
        )
    }

    @PostMapping("/login")
    fun login(@RequestBody request: LoginDto.Request): ResponseEntity<LoginDto.Response> {
        return ResponseEntity.status(HttpStatus.OK).body(
            LoginDto.Response.of(
                accountService.login(request)
            )
        )
    }

    @GetMapping("/info")
    fun userInfo(@RequestParam name: UserInfoDto.Request): ResponseEntity<UserInfoDto.Response> {
        return ResponseEntity.status(HttpStatus.OK).body(
            UserInfoDto.Response.of(
                accountService.getAccountInfo(name.name)
            )
        )
    }

    @GetMapping("/info/all")
    fun userInfoAll(): ResponseEntity<UserInfoDto.BulkResponse> {
        return ResponseEntity.status(HttpStatus.OK).body(
            UserInfoDto.BulkResponse.of(
                accountService.getAccountInfoAll().map { UserInfoDto.Response.of(it) }
            )
        )
    }

    @GetMapping("/admin")
    fun checkAdmin(token: CheckAdminDto.Request): ResponseEntity<CheckAdminDto.Response> {
        return ResponseEntity.status(HttpStatus.OK).body(
            CheckAdminDto.Response.of(
                accountService.isAdmin(token.token)
            )
        )
    }

    @PostMapping("/verify/token")
    fun verifyToken(@RequestBody request: VerifyTokenDto.Request):  ResponseEntity<VerifyTokenDto.Response> {
        val response = accountService.verifyToken(request.tag, request.token)
        return ResponseEntity.status(HttpStatus.OK).body(
            VerifyTokenDto.Response.of(response.status)
        )
    }

    @ResponseBody
    @GetMapping("/profile/tag")
    fun profile(@RequestParam token: ProfileDto.TokenRequest): ResponseEntity<ProfileDto.Response> {
        val tag = accountService.getUserTag(token.token)
        return ResponseEntity.status(HttpStatus.OK).body(
            ProfileDto.Response.of(clanService.getMemberStatisticByTag(tag))
        )
    }
}
